### DISCUSSION TITLE
Should we use SHAP to explain training set or test set?

### DISCUSSION THREAD

**Discussion by rutherz34** [shap/shap#3479] [Q&A]
Hi all,

I have a project that requires predicting preterm births using XGBoost vs. Logistic Regression. I want to use SHAP values to explain the model.

```
# For XGBoost
explainer = shap.TreeExplainer(model)
shap_values = explainer.shap_values(X_train) # Should we use training or test set?

# For Logistic Regression
explainer = shap.Explainer(model, X_train)
shap_values = explainer.shap_values(X_train) # Should we use training or test set?
```

May I ask if the above code is correct, especially the last line of code for each alogrithm? This is because in this GitHub it mentions that we should explain using the test set (i.e., explainer.shap_values(X_test)), however there are certain discussion on Kaggle that mentions using the training set because we are explaining the model, which was formed by the training set?

Furthermore, I would also like to ask if it is appropriate to use mean SHAP values as an approach to feature selection.
